confluent_kafka/schema_registry/config.py

```diff
@@ -0,0 +1,295 @@
+
+            URL(default): Credentials are pulled from the URL if available.
+
+        - schema.registry.basic.auth.user.info (str, optional): Basic auth
+            credentials in the form {username}:{password}.
+        - schema.registry.sasl.username (str, optional): SASL username for use
+            with the PLAIN mechanism.
+        - schema.registry.sasl.password (str, optional): SASL password for use
+            with the PLAIN mechanism.
+        - schema.registry.sasl.mechanism (str, optional): SASL mechanism to
+            use for authentication. Only ``PLAIN`` is supported by
+            ``basic.auth.credentials.source`` SASL_INHERIT.
+
+    """
+    properties = {
+        URL: None,
+        CA_LOCATION: None,
+        CERTIFICATE_LOCATION: None,
+        KEY_LOCATION: None,
+        BASIC_AUTH_CREDENTIALS_SOURCE: BasicAuthCredentialSource.URL,
+        BASIC_AUTH_USER_INFO: None,
+        SASL_USERNAME: None,
+        SASL_PASSWORD: None,
+        SASL_MECHANISM: 'GSSAPI',
+        REST_CLIENT: DefaultRestClient
+    }
+
+    def _err_str(self, val):
+        return self.prefix + val
+
+    def validate(self):
+        # URL validation test requires a string method
+        if not isinstance(self[URL], basestring):
+            raise TypeError("{} must be a string".format(self._err_str(URL)))
+
+        if not self[URL] or not self[URL].startswith("http"):
+            raise ValueError('Invalid URL {}'.format(self[URL]))
+
+        # Ensure we have both keys if set
+        if bool(self[CERTIFICATE_LOCATION]) != bool(self[KEY_LOCATION]):
+            raise ValueError("{} and {} must be set together".format(
+                self._err_str(CERTIFICATE_LOCATION),
+                self._err_str(KEY_LOCATION)))
+
+        if self[BASIC_AUTH_CREDENTIALS_SOURCE] == \
+                BasicAuthCredentialSource.SASL_INHERIT:
+            if bool(self[SASL_USERNAME]) != bool(self[SASL_PASSWORD]):
+                raise ValueError("{} and {} must be set together".format(
+                    self._err_str(SASL_USERNAME),
+                    self._err_str(SASL_PASSWORD)))
+
+            if self[BASIC_AUTH_CREDENTIALS_SOURCE] == \
+                    BasicAuthCredentialSource.SASL_INHERIT and \
+                    'GSSAPI' == self[SASL_MECHANISM]:
+                raise ValueError("{} does not support GSSAPI".format(
+                    self._err_str(BASIC_AUTH_CREDENTIALS_SOURCE)))
+
+        if self[BASIC_AUTH_CREDENTIALS_SOURCE] == \
+                BasicAuthCredentialSource.USERINFO:
+
+            if self[BASIC_AUTH_USER_INFO] is None:
+                raise ValueError("{} is required when {} is {}".format(
+                    self._err_str(BASIC_AUTH_USER_INFO),
+                    self._err_str(BASIC_AUTH_CREDENTIALS_SOURCE),
+                    BasicAuthCredentialSource.USERINFO))
+
+    @property
+    def Url(self):
+        return self[URL]
+
+    @property
+    def SchemaRegistry(self):
```

Comment:
this is upside down, thing that makes an SR, whatever that is in python, should accept a config.

Reply:
I'm, probably wrongly, using the config as a vehicle of dependency injection. It's less for direct instantiation and more for when it's embedded in something like a serializer.